Address David's review (#207)
tfpauly/draft-pauly-adaptive-dns-privacy: draft-pauly-dprive-oblivious-doh.md

## HTTP Request {#oblivious-request}

Oblivious DoH queries are created by the Client, and sent to the Proxy as HTTP
requests using the POST method. Clients are configured with a Proxy URI Template
{{!RFC6570}} and the Target URI. The scheme for both the Proxy URI Template and
the Target URI MUST be "https". The Proxy URI Template contains two variables: 
"targethost", which indicates the host name of the Target server, and "targetpath",
which indicates the path on which the Target's DoH server is running. Examples of

> URI templates are shown below:

Reviewer: ```suggestion
Proxy URI Templates are shown below:
```

which DoH server to use as a Target by specifying two variables: "targethost",
which indicates the host name of the Target server, and "targetpath", which
indicates the path on which the Target's DoH server is running. See

> {{request-example}} for an example request. Clients are configured with a Proxy

Reviewer: I would define the URI template before its variables, and add some restrictions:

```
Oblivious DoH queries are created by the Client, and sent to the Proxy
as an HTTP request using the POST method. Clients are configured with a Proxy
URI Template {{!RFC6570}} which contains two variables: "targethost",
which indicates the host name of the Target server, and "targetpath", which
indicates the path on which the Target's DoH server is running. Examples  of
URI templates are shown below:

~~~
https://dnsproxy.example/dns-query{?targethost,targetpath}
https://dnsproxy.example/{targethost}/{targetpath}
~~~

The URI template MUST contain both the "targethost" and "targetpath" variables
exactly once, and  MUST NOT contain any other variables. The variables MUST be
within the path component of the URI.

See {{request-example}} for an example request.
```

Author: Good suggestion! Fixed.

to indicate that this request is an Oblivious DoH query intended for proxying. Clients
also SHOULD set this same value for the HTTP Accept header.

A correctly encoded request has the HTTP Content-Type header "application/oblivious-dns-message",
uses the HTTP POST method, and contains "targethost" and "targetpath" variables.

> The Proxy MUST validate these parameters and construct a URI Template as follows:

Reviewer: Proxies don't construct a URI template, they just construct an HTTP request

Author: They use the URI Template to generate the request. I tried to clarify.

1. Let $TARGET be the "targethost" parameter from the Client's request, encoded as the

> concatenation of a "host" value (Section 3.2.2 of {{!RFC3986}}) and, optionally, the

Reviewer: This encoding shouldn't be phrased as a Proxy requirement since it applies to both Client and Proxy, I'd move this closer to the URI template definition

Author: Fixed.

":" character followed by port (Section 3.2.3 of {{!RFC3986}}). If this parameter cannot
be encoded as such, the Client's request is incorrectly encoded.

> 1. Let $PATH be the "targetpath" parameter from the Client's request, which MUST be a URI

Reviewer: How is the path encoded? I suspect this requires percent encoding. Also, this isn't a URI template - it's a path, right?

Author: Fixed.

Template {{!RFC6570}}. If this parameter is not a valid URI Template, the Client's request
is incorrectly encoded.

> 1. The Target request HTTPS URI Template is "https://$TARGET$PATH".

Reviewer: I don't understand why a URI is constructed here, just say that the proxy constructs a new request with the host and path it extracted - if you construct a URI then parse again you open yourself up to confusion

Author: Fixed.

to indicate that this request is an Oblivious DoH query intended for proxying. Clients
also SHOULD set this same value for the HTTP Accept header.

A correctly encoded request has the HTTP Content-Type header "application/oblivious-dns-message",

> uses the HTTP POST method, and contains "targethost" and "targetpath" variables. The Proxy constructs 

Reviewer: We should be a bit more pedantic about what `contains "targethost" and "targetpath" variables` means. URI templates don't necessarily translate to path variables. In other words, the requirement is not to contain variables, it's for the proxy to be able to extract them from the path. RFC 6570 seems to call this process "variable matching" so I'd say something along the lines of `If the proxy fails to match the "targethost" and "targetpath" variables from the path, it MUST treat the request as malformed.`

Author: Fixed.

to indicate that this request is an Oblivious DoH query intended for proxying. Clients
also SHOULD set this same value for the HTTP Accept header.

A correctly encoded request has the HTTP Content-Type header "application/oblivious-dns-message",
uses the HTTP POST method, and contains "targethost" and "targetpath" variables. Let

> $TARGET be the "targethost" variable and $PATH be the "targetpath" variable from this

Author: ```suggestion
$TARGET be the "targethost" variable and $PATH be the "targetpath" variable decoded from this
```

> $TARGET be the "targethost" variable and $PATH be the "targetpath" variable from this

Author: ```suggestion
$TARGET be the "targethost" variable and $PATH be the "targetpath" variable percent-decoded from this
```

> A correctly encoded request has the HTTP Content-Type header "application/oblivious-dns-message",
> uses the HTTP POST method, and contains "targethost" and "targetpath" variables. Let
> $TARGET be the "targethost" variable and $PATH be the "targetpath" variable percent-decoded from this
> request. The Proxy constructs a HTTP request to forward the
> Client request to the Target using the URI Template "https://$TARGET$PATH".
> 

Reviewer 2: ```suggestion
A correctly encoded request has the HTTP Content-Type header "application/oblivious-dns-message",
uses the HTTP POST method, and contains "targethost" and "targetpath" variables. The Proxy constructs the URI of the Target with the "https" scheme, using the value of "targethost" as the URI host, and using the value of "targetpath" as the URI path.
```

also SHOULD set this same value for the HTTP Accept header.

A correctly encoded request has the HTTP Content-Type header "application/oblivious-dns-message",
uses the HTTP POST method, and contains "targethost" and "targetpath" variables. Let
$TARGET be the "targethost" variable and $PATH be the "targetpath" variable from this

> request, encoded as described above. The Proxy constructs a HTTP request to forward the

Reviewer: Doesn't the proxy need to percent-decode the path first?

Author: Good catch -- this should read "decoded" from the request.

Author: ```suggestion
request. The Proxy constructs a HTTP request to forward the
```

> Client request to the Target using the URI Template "https://$TARGET$PATH".

Reviewer: I'm not sure what this means. What variables does this URI template contain?

Author: I'm not really sure what the confusion is? $TARGET is "targethost" and $PATH is "targetpath." Is that not specific or clear enough? Or is it not correct?

## HTTP Request Example {#request-example}

The following example shows how a Client requests that a Proxy, "dnsproxy.example",

> forwards an encrypted message to "dnstarget.example". The URI Template for the

Reviewer: This example mentions two URI templates: one for the proxy and one for the target. The normative text above should describe both (it only describes the first one right now). But if I'm reading this document correctly, the "Target URI Template" is not a URI template, it's just a URI. So I'd say in the normative section that the Client is configured with both a Proxy URI Template and a Target URI. (Oh and while we're at it, say that the scheme for both MUST be HTTPS and clients MUST check that)

Author: I _think_ a template without any parameters is still a valid template, but you're right that this is probably confusing. Took the suggestion and fixed. 👍
